src/main/java/software/amazon/cloudformation/Wrapper.java

```diff
@@ -0,0 +1,567 @@
+            this.providerEventsLogger.refreshClient();
+        }
+    }
+
+    public void processRequest(final InputStream inputStream, final OutputStream outputStream) throws IOException,
+        TerminalException {
+
+        ProgressEvent<ResourceT, CallbackT> handlerResponse = null;
+        HandlerRequest<ResourceT, CallbackT> request = null;
+        scrubFiles();
+        try {
+            if (inputStream == null) {
+                throw new TerminalException("No request object received");
+            }
+
+            String input = this.serializer.decompress(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
+
+            JSONObject rawInput = new JSONObject(new JSONTokener(input));
+            // deserialize incoming payload to modelled request
+            try {
+                request = this.serializer.deserialize(input, typeReference);
+            } catch (MismatchedInputException e) {
+                JSONObject resourceSchemaJSONObject = provideResourceSchemaJSONObject();
+                JSONObject rawModelObject = rawInput.getJSONObject("requestData").getJSONObject("resourceProperties");
+                this.validator.validateObject(rawModelObject, resourceSchemaJSONObject);
+            }
+
+            handlerResponse = processInvocation(rawInput, request);
+        } catch (final ValidationException e) {
+            String message;
+            String fullExceptionMessage = ValidationException.buildFullExceptionMessage(e);
+            if (!StringUtils.isEmpty(fullExceptionMessage)) {
+                message = String.format("Model validation failed (%s)", fullExceptionMessage);
+            } else {
+                message = "Model validation failed with unknown cause.";
+            }
+
+            publishExceptionMetric(request == null ? null : request.getAction(), e, HandlerErrorCode.InvalidRequest);
+            handlerResponse = ProgressEvent.defaultFailureHandler(new TerminalException(message, e),
+                HandlerErrorCode.InvalidRequest);
+        } catch (final Throwable e) {
+            // Exceptions are wrapped as a consistent error response to the caller (i.e;
+            // CloudFormation)
+            e.printStackTrace(); // for root causing
+            handlerResponse = ProgressEvent.defaultFailureHandler(e, HandlerErrorCode.InternalFailure);
+            if (request != null && request.getRequestData() != null && MUTATING_ACTIONS.contains(request.getAction())) {
+                handlerResponse.setResourceModel(request.getRequestData().getResourceProperties());
+            }
+            if (request != null) {
+                publishExceptionMetric(request.getAction(), e, HandlerErrorCode.InternalFailure);
+            }
+
+        } finally {
+            if (handlerResponse != null) {
+                publishExceptionCodeAndCountMetric(request == null ? null : request.getAction(), handlerResponse.getErrorCode(),
```

Comment:
There is more new changes on the LabmdaWrapper. Make sure merge the latest change after the refactoring, otherwise it will cause regression.   A recent PR will have new changes on this metric publish.

Reply:
Since I'm refactoring, any changes to the LambdaWrapper will cause merge conflicts so I will do this at the end and make sure the merge is also reviewed. I just don't want to keep fixing the merge conflicts while also making code changes so it's not as confusing looking at the diffs